Add proxy proposal (#2058)
karmada-io/karmada: docs/proposals/resource-aggregation-proxy/README.md

![with-karmada](with-karmada.svg)

We can't access pod and node information from Karmada API server. So after introducing Karmada, the PaaS components can no longer access pod and node resource, this will break the PaaS components. 

In order to solve the problem, we can either:

1. **Modify all of PaaS components.** Force all of them understand Karmada's logic, and find the pod node in the appropriate cluster by themselves.
2. **Provide a Resource Aggregation Proxy.** All PaaS components just ask the proxy to give them the pod node information, the proxy will find the right clusters to get data, then merge data, then return to client.

Apparently, use the proxy approach will greatly simplify external component's access to kubernetes cluster, and hide multiple cluster's implementation details from them. It's what we propose.

Why not use :

- **[Aggregated Kubernetes API Endpoint](https://github.com/karmada-io/karmada/blob/master/docs/userguide/aggregated-api-endpoint.md)** This component is built for Unified authentication, the restful api client still need to enumerate all the member clusters to get the right information.

> - **[Caching](https://github.com/karmada-io/karmada/tree/master/docs/proposals/caching) (aka karmada-search)** This component can only support get/list data, it cannot watch or update data. 

Reviewer: ```suggestion
- **[Caching](https://github.com/karmada-io/karmada/tree/master/docs/proposals/caching) (aka karmada-search)** This component can only get/list, not watch or update data. 
```

# Motivation

> Write a Resource Aggregation Proxy for Karmada. This proxy provide native kubernetes api for components outside kubernetes. When client requests (include read & write operation) proxy, proxy will figure out the correct downstream kubenetes api servers (including Karmada API server and member clusters), request all of the selected servers, then merge the responses into one single response, then return the result to the client. External components can treat Karmada managed kubernetes clusters as one single cluster.

Reviewer: ```suggestion
Write a Resource Aggregation Proxy for Karmada. This proxy provides native kubernetes API for components outside kubernetes. When the client requests operations (including reads & writes), the proxy will figure out the correct downstream kubernetes API servers (including Karmada API server and member clusters), request all the selected servers, and merge the responses into one before responding to the client. External components can treat Karmada-managed kubernetes clusters as one single cluster.
```

## Goals

> Allow kubernetes api clients (kubectl, client-go, and other client use kubernetes api) read write following resources:

Reviewer: ```suggestion
Allow kubernetes API clients (kubectl, client-go, and other clients that use kubernetes APIs) to read/write the following resources:
```

> - workload: create/get/list/watch/update/patch/delete

Author: updated

- node: get/list/watch/update/patch
- pod: get/list/watch/update/patch/logs/exec

## Non-Goals

> - read write following resources:

Reviewer: ```suggestion
- read/write resources:
```

  - service account token

>   - event

Author: updated

> - Provide a solution for unique resource names across all member clusters.

Reviewer 2: Yeah, I get it from below section:

> Resource names need to be unique across all member clusters. Otherwise, the proxy will not be able to differentiate which resource needs to be accessed.

But how to assured that the resource name is unique?

> - Provide a solution for unique resource names across all member clusters.

Author: It is out the scope of this proposal. But i have a rough idea.
We can add a webhook in member clusters, used to validate the resource creations. When validating creations, webhook will search the name from `proxy` component in controller panel.

> - Provide a solution for unique resource names across all member clusters.

Reviewer 2: > We can add a webhook in member clusters, used to validate the resource creations. When validating creations, webhook will search the name from proxy component in controller panel.

Personally, I don't think people would like it. a little bit complicated I guess.

# Proposal

> Add a Resource Aggregation Proxy. When client try to list pod by label, this proxy will aggregate pods information from all member clusters, then return to client as a single list. When client try to get workload resource, proxy will get the information from Karmada API server directly.

Reviewer: ```suggestion
Add a Resource Aggregation Proxy. When the client tries to list pod by label, this proxy will aggregate pods information from all member clusters, and then return to the client as a single list. When the client tries to query workload resources, the proxy will get the information from Karmada API server directly.
```

## User Stories

### Story 1: Service Discovery

> Not everyone use kubernetes Service for service discovery. We use a private service discovery architecture to accommodate both containerized application and traditional application deployed on physical machines. 

Reviewer: ```suggestion
Not everyone uses kubernetes Services for service discovery. We use a private service discovery architecture to accommodate both containerized and traditional applications deployed on physical machines. 
```

> ![user story - service discovery](user story - service discovery.svg)

Reviewer 3: It doesn't seem to be displayed properly here.

![image](https://user-images.githubusercontent.com/78244870/175901040-7b5bf9d4-62c3-42a2-a62c-52615e42d352.png)

Author: Updated

> ![user story - service discovery](user story - service discovery.svg)

Author: ```suggestion
![user story - service discovery](user-story-service discovery.svg)
```

> Its main workflow can be summaried as following:

Reviewer: ```suggestion
Its main workflow can be summarized as follows:
```

> 1. list/watch pods.

Reviewer: ```suggestion
1. List/watch pods.
```

> 2. When pod's ip changed, update the corresponding application's routing metadata.

Reviewer: ```suggestion
2. Update the corresponding application's routing metadata when the pod IP changes.
```

> 3. Patch the pod, add a custom annotation to mark the pod as already processed. 

Reviewer: ```suggestion
3. Patch the pod and mark it as processed by adding a custom annotation. 
```

> The service discovery controller ignore pod with this annotation, to avoid redundant routing metadata change. Workload redeployment or other operation caused pod to be deleted/replaced will automatically remove the added annotation, and make the pod eligible for next processing. 

Reviewer: ```suggestion
The service discovery controller ignores pods with this annotation to avoid redundant routing metadata change. Workload redeployment or other operations that cause pods to be deleted/replaced will automatically remove the added annotation, and make the pods eligible for the next steps of processing. 
```

We need Karmada to provide following functionality:

> 1. list/watch/patch pod

Reviewer 2: Are the Pods restricted to the one that propagated by Karmada? 
In other words, if the user can list pods that they created directly on member clusters?

Author: All the pods.

Reviewer 2: Got it. It makes sense to me.

1. Find a pod in the web UI
2. Click "log" button
3. Pod's log will be displayed in the web UI

We need Karmada to provide following functionality:

1. list all pods
2. access pod's logs

## Notes/Constraints/Caveats

Resource names need to be unique across all member clusters. Otherwise, the proxy will not be able to differentiate which resource needs to be accessed.

# Design Details

## Architecture

Like [Aggregated Kubernetes API Endpoint](https://github.com/karmada-io/karmada/blob/master/docs/userguide/aggregated-api-endpoint.md), Resource Aggregation Proxy will extend kubernetes api by making itself an Aggregated API server.

Client needs to add prefix to the API server url. In our test, Client-go can accept such url during its initiation, and automatically rewrite all request's url with the prefix. Users of client-go do not need to change code in their program. Of course, kubectl can accept the prefixed version of url, too.

In order to accelerate access speed of member cluster resources, we choose to cache resources in proxy.

![Design-Details-Architecture](Design-Details-Architecture.svg)

> The main workflow can be summarized as following:

Reviewer: ```suggestion
The main workflow can be summarized as follows:
```

> 1. Client calls kubernetes native api, the request is sent to Karmada API server with prefixed url, Karmada API server redirect the request to Resource Aggregation Proxy.

Reviewer: ```suggestion
1. Client calls kubernetes native APIs. The request is sent to Karmada API server with the prefixed URL and redirected to Resource Aggregation Proxy.
```

> 2. Resource Aggregation Proxy removes prefix from url, routes the request by resource type:

Reviewer: ```suggestion
2. Resource Aggregation Proxy removes the prefix from the URL and routes the request by resource type:
```

   1. Resource in Karmada control plane, like deployment: Redirect the request to Karmada API server.
   2. Resource in member clusters:

>       1. Read operation, like get/list/watch: Read data from cache, merge the results then return.

Reviewer: ```suggestion
      1. Read operation, like get/list/watch: Read data from cache, merge the results, and then return them.
```

	// Items holds a list of GlobalResource.
	Items []GlobalResource
}

// +k8s:conversion-gen:explicit-from=net/url.Values
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

// ProxyOptions define a flag for proxy that do not have actual resources.
type ProxyOptions struct {
	metav1.TypeMeta `json:",inline"`

	// Path is the part of URLs used for the current proxy request.
	// For example, the whole request URL is
	// http://localhost/apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/nodes
	// Path is api/v1/nodes
	// +optional
	Path string `json:"path,omitempty" protobuf:"bytes,1,opt,name=path"`
}
```

### Example

Users create custom `GlobalResource` as below:

```yaml
apiVersion: proxy.karmada.io/v1alpha1
kind:  GlobalResource
metadata:
  name: proxy-sample
spec:
  targetCluster:
    clusterNames:
    - member1
    - member2
    - member3
  resourceSelectors:
    - apiVersion: v1
      kind: Pod
    - apiVersion: v1
      kind: Node
```

## How to Merge Result

- Get request: return first result.
- List request: concat the results, and return to client as a single list.
- Watch request: return result one by one.

## Resource Version

> Kubernetes uses ResourceVersion to uniquely identify every version of a resource. We merge all clusters' ResourceVersion into one JSON struct, `base64.RawURLEncoding` it, and get the final string.

Reviewer 2: Why user or client needs the `resource version`?

Kubernetes could guarantee the `ResourceVersion` comparable and incremental, can we? or should we?

Author: > Why user or client needs the `resource version`?

client use `resource version` to start a watch.

```
kubectl get po -w --v 6
...
https://192.168.50.83:8443/api/v1/namespaces/default/pods?limit=500 200 OK in 13 milliseconds
...
https://192.168.50.83:8443/api/v1/namespaces/default/pods?resourceVersion=13446488&watch=true 200 OK in 1 milliseconds
...
```

> Kubernetes could guarantee the `ResourceVersion` comparable and incremental, can we? or should we?

Kubernetes's `ResourceVersion` actually is the  ResourceVersion of `etcd`, it is comparable and incremental.

> Kubernetes use ResourceVersion to uniquely identify every version of resource. We merge all clusters' ResourceVersion into one json struct, then `base64.RawURLEncoding` it, get the final string.

Reviewer: ```suggestion
Kubernetes uses ResourceVersion to uniquely identify every version of a resource. We merge all clusters' ResourceVersion into one JSON struct, `base64.RawURLEncoding` it, and get the final string.
```

```
base64.RawURLEncoding({"cluster1":"1234","cluster2":"5678"}) = "eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ"
```

## Example

There are two member clusters: cluster1 and cluster2.

|                 | cluster1         | cluster2         |
| --------------- | ---------------- | ---------------- |

> | resourceVersion | 1234             | 5678             |

Reviewer 2: Is the `1234` the `.metadata.resourceversion` of `Cluster` object?

Reviewer 2: In this table, there are two Pods (`nginx-1` and `nginx-2`), how do you know who is the `latest`?

Author: Indeed, we needn't to care which pod is latest. Response from apiserver will tell it, the field `.metadata.resourceversion` of `PodList`.
```
GET /api/v1/namespaces/test/pods
---
200 OK
Content-Type: application/json

{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {"resourceVersion":"1234"},
  "items": [...]
}
```

Reviewer 2: Oh, good to know. I didn't notice the `List` also has the `resourceVersion`. Thanks for the clarification. 
Do you think we should add it to the proposal? I think it would get easy for other people to understand.

Author: updated

| pods            | nginx-1, nginx-2 | nginx-3, nginx-4 |

**List Pod**

Situation: Client wants to list all pods in cluster1 and cluster2.

1\. Client sends request with prefixed url.

```
GET /apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/namespaces/default/pods
```

2\. Proxy query clusters1's cache, get nginx-1, nginx-2; query cluster2's cache, get nginx-3, nginx-4.

3\. Proxy merges pods' list and returns.

```yaml
apiVersion: v1
kind: List
metadata:
  # eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ is result of 
  # base64.RawURLEncoding("{"cluster1":"1234","cluster2":"5678"}")
  resourceVersion: eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ
items:
- apiVersion: v1
  kind: Pod
  metadata:
    name: nginx-1
  ...
- apiVersion: v1
  kind: Pod
  metadata:
    name: nginx-2
  ...
- apiVersion: v1
  kind: Pod
  metadata:
    name: nginx-3
  ...
- apiVersion: v1
  kind: Pod
  metadata:
    name: nginx-4
  ...
```

**Get Pod**

Situation: Client wants to get Pod by name "nginx-2".

1\. Client sends request with prefixed url.

```
GET /apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/namespaces/default/pods/nginx-2
```

> 2\. Proxy queries all caches, and finds out the required Pod is in cluster1's cache.

Reviewer 2: I know we assume that each pod is unique, but what if two pods with the same name, what's the behavior here?
returns one by random or panic?

Author: Good question. I prefer to return the first one (order by the cluster name). For these reasons:
1. panic is two serious.
2. the returns is stable for client.

Reviewer 2: I agree.
I think ensuring each pod is unique is the biggest challenge for this feature. Do you have any idea to solve it?

3\. Proxy modifies the resourceVersion field then returns response to client.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx-2
  namespace: default
  resourceVersion: eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ
  ...
```

**Watch Pod**

Situation: Client wants to watch Pod.

1\. Client sends request with prefixed url.

```
GET /apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/namespaces/default/pods?resourceVersion=eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ&watch=true
```

2\. Proxy watches cluster1 and cluster2's cache.

> 3\. When there is change, proxy merges the event, modifies the resourceVersion field, then returns response to client.

Reviewer 4: 1. In the watch situation. The proposal will collect the member clusters events and re-allocate global resouceVersion ？
2. Accutally we now use clusterpedia to do `Get` operation for collect member clusters and use  aa to router clusterpedia and karmada.  And we plan to impl watch by `1`.  I think it is same with your proposal. 
Maybe we can help to join your proposal.

Author: > re-allocate global resouceVersion ？

you mean the RV is independent from member RV? In fact, we do not want to maintain this RV. The global RV is built with member RVs in a json. See it in `Resource Version` section.

> 2\. Maybe we can help to join your proposal.

Thanks~

| -------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| url can be the same as kubernetes api server | ⭐️ ✔                                                          | ✘                                                            |
| supported Authentication strategies          | X509 Client Certs                                            | ⭐️ All k8s supported Authentication strategies: X509 Client Certs, Static Token, Bootstrap Token, Service Account Token, OpenID Connect Token, Webhook Token Authentication, Authenticating Proxy |
| performance                                  | ⭐️ Higher. Much shorter api request path: client => proxy => kube-apiserver | Lower. Longer api request path: client => kube-apiserver => proxy => kube-apiserver |
| developerment complexity                     | High. In order to support all kubernetes Authentication strategies, we need to include kubernetes Authentication module's code, and update proxy when kubernetes updates. | ⭐️ Low. Only need to write an aggregated API server.          |

## Conclusion

> Some users may need to use Authentication strategies other than X509 Client Certs, but implement it in proxy itself will be too complex. So we consider putting the proxy behind API server a better option.

Reviewer 2: > So we consider putting the proxy behind API server a better option.

Does that means some changes are needed from `PaaS components`?

Author: The only changes for `PaaS components` is change the url of `apiserver`.
Originally, it may be
```
server=http://10.10.10.10:5443
```

Then  change it to
```
server=http://10.10.10.10:5443/apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy
```

Reviewer 2: em, both the API and implementation are similar to karmada-search now. 
Have you ever thought about putting this feature into karmada search?

Author: I have read the `karmada search` codes, it uses `Informer` to cache resource. It is difficult and complex to implement
[Semantics of resourveVersion](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list), [bookmarks](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks), [resourceVersionMatch](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list)

In `proxy` component, i will use [k8s.io/apiserver/pkg/storage/cacher.Cacher](https://github.com/kubernetes/kubernetes/blob/340555476a4926e10c9ffac8b283cd77ca088c5d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L226-L296), it is native to support all scenes of get/list/watch

Reviewer 2: It's not my preference, just thinking about it.

For `proxy` purposes, we also use `informer` to collect resources from member clusters, right? Is it easy for karmada-search using the `k8s.io/apiserver/pkg/storage/cacher.Cacher`?
cc @liys87x here

Author: > For `proxy` purposes, we also use `informer` to collect resources from member clusters, right? I

No. `cacher.Cacher` also list/watch from member clusters, collecting resources.

If `karmada-search` use it, is shall start `informer` from the `cacher.Cacher`, instead of member cluster apiserver.

@liys87x please evaluate it.

Reviewer 5: I found that `cacher.Cacher` is embedded in restOption and I am learning how it is used.

Reviewer 5: Then I will evaluate it.

Reviewer 6: so the karmada aggregation proxy will :
1. start informers listening to each member cluster and watch registered resources from member clusters
2. watched resources and theire changes from the member cluster will be stored into some etcd storage
3. start a cacher.Cacher with backend storage pointing to etcd storage above and serve list/watch requests

is my understanding correct?  @ikaven1024

Author: There is no informers and etcd in proxy. Indeed, the backend storage of `Cacher` is `kube-apiserver` in member clusters. Cacher list & watch from member apiserver, store member resources in memory.

Author: Store all resources from member cluster, it is a big challenge to etcd. So we don't use etcd in proxy.

Reviewer 6: > 

thank you for your kind explanation @ikaven1024 
two more question please :  
    if cacher.Cacher is used , we will have n cacher.Cacher object for n member clusters ? 
    when a watch request is served, we will have a aggregation proxy to start watch from each cacher.Cacher object, then merge the events from each cacher.Cacher, modify resourceVersion, finally return values to the user?

Author: > we will have n cacher.Cacher object for n member clusters

Right

> when a watch request is served, we will have a aggregation proxy to start watch from each cacher.Cacher object, then merge the events from each cacher.Cacher, modify resourceVersion, finally return values to the user?

Right

> Some users may need to use Authentication strategies other than X509 Client Certs, but implement it in proxy itself will be too complex. So we consider put proxy behind API server will be a better option.

Reviewer: ```suggestion
Some users may need to use Authentication strategies other than X509 Client Certs, but implement it in proxy itself will be too complex. So we consider putting the proxy behind API server a better option.
```
